src/meshTools/AMIInterpolation/AMIInterpolation/AMIMethod/directAMI/directAMI.C

```diff
@@ -24,6 +24,82 @@ License
 \*---------------------------------------------------------------------------*/
 
 #include "directAMI.H"
+#include "primitiveFieldsFwd.H"
+#include "pointFieldFwd.H"
+#include "faceData.H"
+#include "faceFunctors.H"
+#include "gpuList.H"
+#include "scalarList.H"
+
+namespace Foam {
+    struct ShootRayFunctor {
+        const point *tgtPoints;
+        const point *srcPoints;
+        const vector *srcCf;
+        const faceData *tgtFaces;
+        const faceData *srcFaces;
+        const label size;
+
+        // Arrays used for addressing
+        label* srcAddr;
+        label* tgtAddr;
+
+        // Arrays used to hold the actual count
+        label* srcCount;
+        label* tgtCount;
+
+        faceRayFunctor rayFunctor;
+        faceNormalFunctor normalFunctor;
+
+        ShootRayFunctor
+        (
+            const point *_tgtPoints,
+            const point *_srcPoints,
+            const vector *_srcCf,
+            const faceData *_tgtFaces,
+            const faceData *_srcFaces,
+            const label _size,
+            label* _srcCount,
+            label* _tgtCount,
+            label* _srcAddr,
+            label* _tgtAddr
+        ) :
+            tgtPoints(_tgtPoints),
+            srcPoints(_srcPoints),
+            srcCf(_srcCf),
+            tgtFaces(_tgtFaces),
+            srcFaces(_srcFaces),
+            size(_size),
+            srcCount(_srcCount),
+            tgtCount(_tgtCount),
+            srcAddr(_srcAddr),
+            tgtAddr(_tgtAddr),
+            rayFunctor(_tgtPoints, intersection::FULL_RAY, intersection::VECTOR),
+            normalFunctor(_srcPoints)
+        {}
+
+        template <typename Tuple>
+        __host__ __device__
+        void operator()(Tuple t)
+        {
+            const label tgtI = thrust::get<0>(t);
+            const label srcI = thrust::get<1>(t);
+
+            if (rayFunctor(tgtFaces[tgtI], srcCf[srcI], normalFunctor(srcFaces[srcI])).hit()) {
+                if (srcCount[srcI] < size)
```

Comment:
Read access is not a problem, when multiple threads want to write to a location, then we have a problem.